js/app/demo/messages.jsx

```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+define(['react', 'jsx!message'], function(React, Message) {
+    return React.createClass({
+        displayName: 'Messages',
+        render: function() {
+            if (!this.props.values || this.props.values.length === 0) {
+                return (
+                    <div id="results">
+                        <div className="alert alert-success">Lint-free!</div>
+                    </div>
+                )
+            } else {
+                return (
+                    <div id="results">
+                        {
+                            this.props.values.map(function(message, index) {
+                                return <Message key={index} value={message} />;
```

Comment:
`key`s are hard in React. For a long time, I used `index` too, which silenced the warning, but didn't actually solve the problem. `key` should be unique (and constant) across renders, so perhaps something like the concatenation of line, column, and rule ID? I'll try to explain via example.

1. Paste the following code in the demo editor:

    ```js
    alert(1);
    alert(2);
    ```

2. You'll see two messages, as expected:

    ![screenshot 2016-04-11 11 31 34](https://cloud.githubusercontent.com/assets/1709537/14432638/f834f996-ffd8-11e5-9dc8-032c03e35bbf.png)

3. Right click -> Inspect on the rule link in the second message. In the dev tools, the `<a>` element should be highlighted.

4. Comment out the first line:

    ```js
    alert(1);
    alert(2);
    ```

5. The first message goes away, and the second *appears* to remain, but the `<a>` is no longer selected in the element inspector. Because the `key` changed, React had to re-render the entire DOM for messages. Ideally, React would be able to detect that the second message is unchanged and wouldn't have to re-render it.

6. Try giving the messages a unique `key`:

    ```diff
    diff --git a/js/app/demo/messages.jsx b/js/app/demo/messages.jsx
    index 40a78fa..ee8b552 100644
    --- a/js/app/demo/messages.jsx
    +++ b/js/app/demo/messages.jsx
    @@ -14,8 +14,10 @@ define(['react', 'jsx!message'], function(React, Message) {
                     return (
                         <div id="results">
                             {
    -                            this.props.values.map(function(message, index) {
    -                                return <Message key={index} value={message} />;
    +                            this.props.values.map(function(message) {
    +                                var key = '' + message.line + ':' + message.column + ':' + message.ruleId;
    +
    +                                return <Message key={key} value={message} />;
                                 })
                             }
                         </div>
    ```

7. Repeat steps 1-4, but this time, rather than re-rendering every message, React only had to remove the first message's DOM element. The `<a>` remained selected. In fact, you can uncomment the first line again, and React will insert a DOM element at the beginning of the message list, but it won't touch the DOM for the message on the second line.